doc/kb_sdk_impl_methods.md

```diff
@@ -0,0 +1,201 @@
+
+Please see:
+
+    https://github.com/kbase/kb_sdk/blob/master/doc/kb_sdk_data_types.md
+
+[\[Back to Edit Impl\]](#impl)
+
+
+##### <A NAME="impl-logging"></A>5.3. Logging
+
+Logging where you are is key to tracking progress and debugging.  Our recommended style is to log to a "console" list.  Here is some example code for accomplishing this.
+
+```python
+    from pprint import pprint, pformat
+
+    # target is a list for collecting log messages
+    def log(self, target, message):
+        # we should do something better here...
+        if target is not None:
+            target.append(message)
+        print(message)
+        sys.stdout.flush()
+
+    def run_<MyMethod>(self, ctx, params):
+        console = []
+        self.log(console,'Running run_<MyMethod> with params=')
+        self.log(console, pformat(params))
+```    
+
+[\[Back to Edit Impl\]](#impl)
+
+##### <A NAME="impl-provenance"></A>5.4. Provenance
+
+Data objects in KBase contain provenance (historical information of their creation and objects from which they are derived).  When you create new objects, you must carry forward and add provenance information to them.  Additionally, Report objects should receive that provenance data (see below).  Examples of adding provenance to objects can be found in the [KBase Data Types](docs/kb_sdk_data_types.md).
+
+```python
+        # load the method provenance from the context object
+        provenance = [{}]
+        if 'provenance' in ctx:
+            provenance = ctx['provenance']
+        # add additional info to provenance here, in this case the input data object reference
+        provenance[0]['input_ws_objects']=[]
+        provenance[0]['input_ws_objects'].append(params['workspace_name']+'/'+params['read_library_name'])
+```
+
+[\[Back to Edit Impl\]](#impl)
+
+##### <A NAME="impl-report"></A>5.5. Building Output Report
+
+```python
+        # create a Report
+        report = ''
+        report += 'ContigSet saved to: '+params['workspace_name']+'/'+params['output_contigset_name']+'\n'
+        report += 'Assembled into '+str(len(contigset_data['contigs'])) + ' contigs.\n'
+        report += 'Avg Length: '+str(sum(lengths)/float(len(lengths))) + ' bp.\n'
+
+        # compute a simple contig length distribution
+        bins = 10
+        counts, edges = np.histogram(lengths, bins)
+        report += 'Contig Length Distribution (# of contigs -- min to max basepairs):\n'
+        for c in range(bins):
+            report += '   '+str(counts[c]) + '\t--\t' + str(edges[c]) + ' to ' + str(edges[c+1]) + ' bp\n'
+
+        reportObj = {
+            'objects_created':[{'ref':params['workspace_name']+'/'+params['output_contigset_name'], 'description':'Assembled contigs'}],
+            'text_message':report
+        }
+
+        reportName = 'megahit_report_'+str(uuid.uuid4())   # uuid4() instead of getnode() to ensure uniqueness
+        report_obj_info = ws.save_objects({
```

Comment:
Wait, KBaseReports has a way of saving items directly to the workspace? One of these sections should probably just be a detailed explanation of KBaseReports because I clearly never came to understand it's full potential.